### PR TITLE
fix: helm rollback in case of no gitops

### PR DESCRIPTION
**Pull request by RajeevRanjan27**
# Description

Fixes https://github.com/devtron-labs/devops-sprint/issues/1011

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code.
* [x] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/appStore/installedApp/service/AppStoreDeploymentService.go [hunk 1]
```diff
@@ -492,12 +492,50 @@ func (impl *AppStoreDeploymentServiceImpl) RollbackApplication(ctx context.Conte
...
 	var success bool
-	if util2.IsHelmApp(installedApp.AppOfferingMode) {
+	//case when the request is for external app case
+	if installedApp.Id == 0 {
 		installedApp, success, err = impl.eaModeDeploymentService.RollbackRelease(ctx, installedApp, request.GetVersion(), tx)
 		if err != nil {
 			impl.logger.Errorw("error while rollback helm release", "error", err)
 			return false, err
 		}
+	} else if installedApp.DeploymentAppType == util.PIPELINE_DEPLOYMENT_TYPE_HELM && installedApp.Id > 0 {
```
Comment (Ash-exp) [inline pkg/appStore/installedApp/service/AppStoreDeploymentService.go:502]: use util.IsHelmApp(deploymentConfig.DeploymentAppType) and remove unnecessary check installedApp.Id > 0

File: pkg/appStore/installedApp/service/AppStoreDeploymentService.go [hunk 2]
```diff
@@ -492,12 +492,50 @@ func (impl *AppStoreDeploymentServiceImpl) RollbackApplication(ctx context.Conte
 	}
 	// Rollback starts
 	var success bool
-	if util2.IsHelmApp(installedApp.AppOfferingMode) {
+	//case when the request is for external app case
+	if installedApp.Id == 0 {
```
Comment (Ash-exp) [inline pkg/appStore/installedApp/service/AppStoreDeploymentService.go:496]: make a common const for this. has been evaluated multiple times

File: pkg/appStore/installedApp/service/AppStoreDeploymentService.go [hunk 3]
```diff
@@ -492,12 +492,50 @@ func (impl *AppStoreDeploymentServiceImpl) RollbackApplication(ctx context.Conte
...
 		installedApp, success, err = impl.eaModeDeploymentService.RollbackRelease(ctx, installedApp, request.GetVersion(), tx)
 		if err != nil {
 			impl.logger.Errorw("error while rollback helm release", "error", err)
 			return false, err
 		}
+	} else if installedApp.DeploymentAppType == util.PIPELINE_DEPLOYMENT_TYPE_HELM && installedApp.Id > 0 {
+		upgradeRequest := &appStoreBean.InstallAppVersionDTO{
+			UserId: userId,
```
Comment (Ash-exp) [inline pkg/appStore/installedApp/service/AppStoreDeploymentService.go:504]: use builder patterns to populate the DTO

File: pkg/appStore/installedApp/service/AppStoreDeploymentService.go [hunk 4]
```diff
@@ -492,12 +492,50 @@ func (impl *AppStoreDeploymentServiceImpl) RollbackApplication(ctx context.Conte
...
+		}
+		upgradeRequest.InstalledAppId = installedApp.InstalledAppId
+		installedAppVersionHistory, err := impl.installedAppRepositoryHistory.GetInstalledAppVersionHistory(int(request.GetVersion()))
+		if err != nil {
+			impl.logger.Errorw("error while fetching installed app version history from Db", "request", request, "error", err)
+			return false, err
+		}
+		if installedAppVersionHistory.Status == "Healthy" || installedAppVersionHistory.Status == "Succeeded" {
```
Comment (Ash-exp) [inline pkg/appStore/installedApp/service/AppStoreDeploymentService.go:512]: bind a method to installedAppVersionHistory. IsSuccessful()

**Review by Ash-exp — CHANGES_REQUESTED** [review]

File: pkg/appStore/installedApp/service/EAMode/InstalledAppDBService.go [hunk 5]
```diff
@@ -51,6 +51,7 @@ type InstalledAppDBService interface {
 	GetInstalledAppByClusterNamespaceAndName(appIdentifier *helmBean.AppIdentifier) (*appStoreBean.InstallAppVersionDTO, error)
 	GetInstalledAppByInstalledAppId(installedAppId int) (*appStoreBean.InstallAppVersionDTO, error)
 	GetInstalledAppVersion(id int, userId int32) (*appStoreBean.InstallAppVersionDTO, error)
+	GetInstalledAppVersionAny(id int, userId int32) (*appStoreBean.InstallAppVersionDTO, error)
```
Comment (Ash-exp) [inline pkg/appStore/installedApp/service/EAMode/InstalledAppDBService.go:54]: rename to `GetInstalledAppVersionByIdIncludeDeleted`

**Review by Ash-exp — CHANGES_REQUESTED** [review]

**Review by Ash-exp — APPROVED** [review]

File: pkg/appStore/installedApp/service/AppStoreDeploymentService.go [hunk 6]
```diff
@@ -456,90 +463,51 @@ func (impl *AppStoreDeploymentServiceImpl) UpdateProjectHelmApp(updateAppRequest
...
-		installedAppVersion.UpdatedBy = userId
-		_, err = impl.installedAppRepository.UpdateInstalledAppVersion(installedAppVersion, tx)
+
+		adapter.GenerateUpdateRequestDTO(upgradeRequest, installedApp, installedAppVersionHistory, installedAppVersionDTO)
+
+		// installedApp.AppStoreVersion is the current chart version
+		// previous state of installedAppVersion.AppStoreVersion
+		if installedApp.AppStoreVersion != installedAppVersionDTO.AppStoreVersion {
+			// indicates that the chart/ chart version has been changed
+			upgradeRequest.Id = 0
```
Comment (iamayushm) [inline pkg/appStore/installedApp/service/AppStoreDeploymentService.go:497]: id should be zero only when chart is changed and not version, verify

File: pkg/appStore/adapter/Adapter.go [hunk 7]
```diff
@@ -113,6 +113,16 @@ func NewInstalledAppDeploymentAction(deploymentAppType string) *appStoreBean.Ins
 	}
 	return installedAppDeploymentAction
 }
+func GenerateUpdateRequestDTO(request *appStoreBean.InstallAppVersionDTO, installedApp *appStoreBean.InstallAppVersionDTO, installedAppVersionHistory *repository.InstalledAppVersionHistory, installedAppVersionDTO *appStoreBean.InstallAppVersionDTO) {
```
Comment (iamayushm) [inline pkg/appStore/adapter/Adapter.go:116]: can be renamed to updateRequestDTOForRollback or something similar

File: pkg/appStore/installedApp/service/AppStoreDeploymentService.go [hunk 8]
```diff
@@ -456,90 +463,51 @@ func (impl *AppStoreDeploymentServiceImpl) UpdateProjectHelmApp(updateAppRequest
...
-		installedAppVersion.ValuesYaml = installedApp.ValuesOverrideYaml
-		installedAppVersion.UpdatedOn = time.Now()
-		installedAppVersion.UpdatedBy = userId
-		_, err = impl.installedAppRepository.UpdateInstalledAppVersion(installedAppVersion, tx)
+
+		adapter.GenerateUpdateRequestDTO(upgradeRequest, installedApp, installedAppVersionHistory, installedAppVersionDTO)
+
+		// installedApp.AppStoreVersion is the current chart version
+		// previous state of installedAppVersion.AppStoreVersion
+		if installedApp.AppStoreVersion != installedAppVersionDTO.AppStoreVersion {
```
Comment (iamayushm) [inline pkg/appStore/installedApp/service/AppStoreDeploymentService.go:495]: this logic can be moved to GenerateUpdateRequestDTO

File: pkg/appStore/installedApp/service/AppStoreDeploymentService.go [hunk 9]
```diff
@@ -851,8 +825,15 @@ func (impl *AppStoreDeploymentServiceImpl) UpdateInstalledApp(ctx context.Contex
...
 		}
 	}
+	err1 := impl.UpdatePreviousDeploymentStatusForAppStore(upgradeAppRequest, triggeredAt, err)
+	if err1 != nil {
+		impl.logger.Errorw("error while update previous installed app version history", "err", err, "installAppVersionRequest", upgradeAppRequest)
+		//if installed app is updated and error is in updating previous deployment status, then don't block user, just show error.
+	}
 
-	return upgradeAppRequest, nil
+	err = impl.attributesService.UpdateKeyValueByOne(bean2.HELM_APP_UPDATE_COUNTER)
```
Comment (iamayushm) [inline pkg/appStore/installedApp/service/AppStoreDeploymentService.go:834]: this value should be update irrespective of update is success/failure, move this back to rest handler

**Review by iamayushm — APPROVED** [review]

**Review by iamayushm — COMMENTED** [review]
please check rollback for apps installed in EA mode, that should we also modify this check "util2.IsHelmApp(installedApp.AppOfferingMode)" in GetDeploymentHistory function to installedAppId>0 . did we checked this case in QA?

**Comment by RajeevRanjan27** [comment]
> please check rollback for apps installed in EA mode, that should we also modify this check "util2.IsHelmApp(installedApp.AppOfferingMode)" in GetDeploymentHistory function to installedAppId>0 . did we checked this case in QA?

> for ea mode rollback we have handled the case inside the `GetDeploymentHistory` function with `installedAppId>0`

**Review by iamayushm — APPROVED** [review]

**Review by vikramdevtron — APPROVED** [review]

**Pull request merged**